[Consumer] Stopped when Kafka broker restarts

Question.
I'm using the handy consumer.run() eachMessage() feature in a Node.js daemon that consumes Kafka messages from one topic continuously. The goal is that the daemon always keeps running, even in the event of failures.

When the Consumer is already connected to the Kafka broker, and the Kafka broker goes down or the network is offline temporarily, then the Consumer keeps running (emitting logs to indicate that it cannot connect) and that is as expected.

But the consumer stops when the Kafka broker (just one) is restarted. The final message is ""[Consumer] Crash: KafkaJSProtocolError: The coordinator is loading and hence can't process requests for this group". No exceptions are thrown so I cannot handle that situation.

=> Question: what is the best method to ensure that the Consumer remains active after restarting the Kafka broker?

```
Code:
    await consumer.run({
        eachMessage: async ({ topic, partition, message }) => {
            const prefix = `${topic}[part ${partition} | offset ${message.offset}] / ${message.timestamp}`;
            console.log(`- ${prefix} ${message.key}#${message.value}`);
        }
    });
```

```
logs:
{"level":"ERROR","timestamp":"2019-02-07T10:53:42.601Z","logger":"kafkajs","message":"[Connection] Connection error: connect ECONNREFUSED 192.168.0.80:9092","broker":"myvmbigdata:9092","clientId":"my-nodejs-playground","stack":"Error: connect ECONNREFUSED 192.168.0.80:9092\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"}
{"level":"ERROR","timestamp":"2019-02-07T10:53:43.929Z","logger":"kafkajs","message":"[Connection] Connection timeout","broker":"myvmbigdata:9092","clientId":"my-nodejs-playground"}
{"level":"ERROR","timestamp":"2019-02-07T10:53:43.929Z","logger":"kafkajs","message":"[BrokerPool] KafkaJSConnectionError: Connection timeout","retryCount":1,"retryTime":656}
{"level":"ERROR","timestamp":"2019-02-07T10:53:51.811Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to seed broker, trying another broker from the list: Connection timeout","retryCount":4,"retryTime":5334}
{"level":"ERROR","timestamp":"2019-02-07T10:53:51.827Z","logger":"kafkajs","message":"[Connection] Connection error: connect ECONNREFUSED 192.168.0.80:9092","broker":"myvmbigdata:9092","clientId":"my-nodejs-playground","stack":"Error: connect ECONNREFUSED 192.168.0.80:9092\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1113:14)"}

{"level":"ERROR","timestamp":"2019-02-07T10:53:57.477Z","logger":"kafkajs","message":"[Connection] Response JoinGroup(key: 11, version: 0)","broker":"myvmbigdata:9092","clientId":"my-nodejs-playground","error":"The coordinator is loading and hence can't process requests for this group","correlationId":3,"size":20}
{"level":"ERROR","timestamp":"2019-02-07T10:53:57.477Z","logger":"kafkajs","message":"[Consumer] Crash: KafkaJSProtocolError: The coordinator is loading and hence can't process requests for this group","groupId":"group-my-nodejs-playground","stack":"KafkaJSProtocolError: The coordinator is loading and hence can't process requests for this group\n    at createErrorFromCode (C:\\myhtdocs\\my-nodejs\\my-nodejs-playground\\node_modules\\kafkajs\\src\\protocol\\error.js:472:10)\n    at Object.parse (C:\\myhtdocs\\my-nodejs\\my-nodejs-playground\\node_modules\\kafkajs\\src\\protocol\\requests\\joinGroup\\v0\\response.js:37:11)\n    at Connection.send (C:\\myhtdocs\\my-nodejs\\my-nodejs-playground\\node_modules\\kafkajs\\src\\network\\connection.js:272:35)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"}
{"level":"INFO","timestamp":"2019-02-07T10:53:57.492Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"group-my-nodejs-playground"}
```

Hi @pantaluna, the expectation is that the consumers will never stop, we use them with this assumption. I performed some tests using the example we have in the repo and they worked well, I'm guessing that we are missing this corner case (`The coordinator is loading and hence can't process requests for this group`). I will create an issue to handle this scenario.

My tests:

> Booting the consumer
```sh
→ node examples/consumer.js
{"level":"INFO","timestamp":"2019-02-07T22:19:43.968Z","logger":"kafkajs","message":"[Consumer] Starting","groupId":"test-group"}
{"level":"INFO","timestamp":"2019-02-07T22:19:47.108Z","logger":"kafkajs","message":"[Runner] Consumer has joined the group","groupId":"test-group","memberId":"example-consumer-a2d5e474-467e-4403-8cc8-88577563c941","leaderId":"example-consumer-a2d5e474-467e-4403-8cc8-88577563c941","isLeader":true,"memberAssignment":{"topic-test":[2,5,4,1,3,0]},"groupProtocol":"RoundRobinAssigner","duration":3139}
```

> Killing one node
```sh
{"level":"ERROR","timestamp":"2019-02-07T22:20:02.285Z","logger":"kafkajs","message":"[Connection] Connection error: socket hang up","broker":"192.168.1.172:9094","clientId":"example-consumer","stack":"Error: socket hang up\n    at TLSSocket.onHangUp (_tls_wrap.js:1137:19)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:111:20)\n    at TLSSocket.emit (events.js:208:7)\n    at endReadableNT (_stream_readable.js:1064:12)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickCallback (internal/process/next_tick.js:180:9)"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:02.285Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to broker, reconnecting","retryCount":0,"retryTime":339}
{"level":"ERROR","timestamp":"2019-02-07T22:20:02.630Z","logger":"kafkajs","message":"[Connection] Connection error: socket hang up","broker":"192.168.1.172:9094","clientId":"example-consumer","stack":"Error: socket hang up\n    at TLSSocket.onHangUp (_tls_wrap.js:1137:19)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:111:20)\n    at TLSSocket.emit (events.js:208:7)\n    at endReadableNT (_stream_readable.js:1064:12)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickCallback (internal/process/next_tick.js:180:9)"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:02.631Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to broker, reconnecting","retryCount":1,"retryTime":546}
{"level":"ERROR","timestamp":"2019-02-07T22:20:03.185Z","logger":"kafkajs","message":"[Connection] Connection error: socket hang up","broker":"192.168.1.172:9094","clientId":"example-consumer","stack":"Error: socket hang up\n    at TLSSocket.onHangUp (_tls_wrap.js:1137:19)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:111:20)\n    at TLSSocket.emit (events.js:208:7)\n    at endReadableNT (_stream_readable.js:1064:12)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickCallback (internal/process/next_tick.js:180:9)"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:03.185Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to broker, reconnecting","retryCount":2,"retryTime":920}
{"level":"ERROR","timestamp":"2019-02-07T22:20:04.113Z","logger":"kafkajs","message":"[Connection] Connection error: socket hang up","broker":"192.168.1.172:9094","clientId":"example-consumer","stack":"Error: socket hang up\n    at TLSSocket.onHangUp (_tls_wrap.js:1137:19)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:111:20)\n    at TLSSocket.emit (events.js:208:7)\n    at endReadableNT (_stream_readable.js:1064:12)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickCallback (internal/process/next_tick.js:180:9)"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:04.113Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to broker, reconnecting","retryCount":3,"retryTime":2022}
{"level":"ERROR","timestamp":"2019-02-07T22:20:07.141Z","logger":"kafkajs","message":"[Connection] Connection timeout","broker":"192.168.1.172:9094","clientId":"example-consumer"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:07.142Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to broker, reconnecting","retryCount":4,"retryTime":4182}
{"level":"ERROR","timestamp":"2019-02-07T22:20:12.331Z","logger":"kafkajs","message":"[Connection] Connection timeout","broker":"192.168.1.172:9094","clientId":"example-consumer"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:12.331Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to broker, reconnecting","retryCount":5,"retryTime":7410}
{"level":"ERROR","timestamp":"2019-02-07T22:20:12.332Z","logger":"kafkajs","message":"[Consumer] Crash: KafkaJSNumberOfRetriesExceeded: Connection timeout","groupId":"test-group","retryCount":1,"stack":"KafkaJSNumberOfRetriesExceeded\n  Caused by: KafkaJSNumberOfRetriesExceeded\n  Caused by: KafkaJSConnectionError: Connection timeout\n    at Timeout.onTimeout [as _onTimeout] (/Users/tulio.ornelas/Projects/github/kafkajs/src/network/connection.js:145:23)\n    at ontimeout (timers.js:498:11)\n    at tryOnTimeout (timers.js:323:5)\n    at Timer.listOnTimeout (timers.js:290:5)"}
{"level":"INFO","timestamp":"2019-02-07T22:20:12.341Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"test-group"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:12.342Z","logger":"kafkajs","message":"[Consumer] Restarting the consumer in 714ms","retryCount":1,"groupId":"test-group"}
{"level":"INFO","timestamp":"2019-02-07T22:20:13.061Z","logger":"kafkajs","message":"[Consumer] Starting","groupId":"test-group"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:14.067Z","logger":"kafkajs","message":"[Connection] Connection timeout","broker":"192.168.1.172:9094","clientId":"example-consumer"}
{"level":"ERROR","timestamp":"2019-02-07T22:20:14.068Z","logger":"kafkajs","message":"[BrokerPool] KafkaJSConnectionError: Connection timeout","retryCount":0,"retryTime":309}
{"level":"ERROR","timestamp":"2019-02-07T22:20:14.068Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to seed broker, trying another broker from the list: Connection timeout","retryCount":0,"retryTime":309}
{"level":"ERROR","timestamp":"2019-02-07T22:20:14.433Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9097","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":3,"size":261}
```

> All nodes down
```sh
{"level":"ERROR","timestamp":"2019-02-07T22:23:46.621Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to seed broker, trying another broker from the list: Connection timeout","retryCount":0,"retryTime":288}
{"level":"ERROR","timestamp":"2019-02-07T22:23:47.454Z","logger":"kafkajs","message":"[Connection] Connection error: socket hang up","broker":"192.168.1.172:9094","clientId":"example-consumer","stack":"Error: socket hang up\n    at TLSSocket.onHangUp (_tls_wrap.js:1137:19)\n    at Object.onceWrapper (events.js:313:30)\n    at emitNone (events.js:111:20)\n    at TLSSocket.emit (events.js:208:7)\n    at endReadableNT (_stream_readable.js:1064:12)\n    at _combinedTickCallback (internal/process/next_tick.js:138:11)\n    at process._tickCallback (internal/process/next_tick.js:180:9)"}
{"level":"ERROR","timestamp":"2019-02-07T22:23:47.817Z","logger":"kafkajs","message":"[Connection] Connection error: connect ETIMEDOUT 192.168.1.172:9100","broker":"192.168.1.172:9100","clientId":"example-consumer","stack":"Error: connect ETIMEDOUT 192.168.1.172:9100\n    at Object._errnoException (util.js:992:11)\n    at _exceptionWithHostPort (util.js:1014:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1186:14)"}
{"level":"ERROR","timestamp":"2019-02-07T22:23:47.913Z","logger":"kafkajs","message":"[Connection] Connection timeout","broker":"192.168.1.172:9100","clientId":"example-consumer"}
{"level":"ERROR","timestamp":"2019-02-07T22:23:47.914Z","logger":"kafkajs","message":"[BrokerPool] KafkaJSConnectionError: Connection timeout","retryCount":1,"retryTime":498}
{"level":"ERROR","timestamp":"2019-02-07T22:23:47.914Z","logger":"kafkajs","message":"[BrokerPool] Failed to connect to seed broker, trying another broker from the list: Connection timeout","retryCount":1,"retryTime":498}
```
It kept trying to reconnect

> With all nodes up again
```sh
{"level":"ERROR","timestamp":"2019-02-07T22:24:15.570Z","logger":"kafkajs","message":"[Connection] Connection error: connect ETIMEDOUT 192.168.1.172:9097","broker":"192.168.1.172:9097","clientId":"example-consumer","stack":"Error: connect ETIMEDOUT 192.168.1.172:9097\n    at Object._errnoException (util.js:992:11)\n    at _exceptionWithHostPort (util.js:1014:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1186:14)"}
{"level":"ERROR","timestamp":"2019-02-07T22:24:16.918Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9094","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":15,"size":228}
{"level":"ERROR","timestamp":"2019-02-07T22:24:16.919Z","logger":"kafkajs","message":"[Consumer] Crash: KafkaJSNumberOfRetriesExceeded: There is no leader for this topic-partition as we are in the middle of a leadership election","groupId":"test-group","retryCount":0,"stack":"KafkaJSNumberOfRetriesExceeded\n  Caused by: KafkaJSNumberOfRetriesExceeded\n  Caused by: KafkaJSProtocolError: There is no leader for this topic-partition as we are in the middle of a leadership election\n    at createErrorFromCode (/Users/tulio.ornelas/Projects/github/kafkajs/src/protocol/error.js:501:10)\n    at Object.parse (/Users/tulio.ornelas/Projects/github/kafkajs/src/protocol/requests/metadata/v0/response.js:66:11)\n    at Connection.send (/Users/tulio.ornelas/Projects/github/kafkajs/src/network/connection.js:286:35)\n    at <anonymous>\n    at process._tickCallback (internal/process/next_tick.js:188:7)"}
{"level":"INFO","timestamp":"2019-02-07T22:24:16.919Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"test-group"}
{"level":"ERROR","timestamp":"2019-02-07T22:24:16.919Z","logger":"kafkajs","message":"[Consumer] Restarting the consumer in 282ms","retryCount":0,"groupId":"test-group"}
{"level":"INFO","timestamp":"2019-02-07T22:24:17.205Z","logger":"kafkajs","message":"[Consumer] Starting","groupId":"test-group"}
{"level":"ERROR","timestamp":"2019-02-07T22:24:17.256Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9094","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":17,"size":228}
{"level":"ERROR","timestamp":"2019-02-07T22:24:17.594Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9094","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":18,"size":228}
{"level":"ERROR","timestamp":"2019-02-07T22:24:18.381Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9094","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":19,"size":228}
{"level":"ERROR","timestamp":"2019-02-07T22:24:18.443Z","logger":"kafkajs","message":"[Connection] Connection error: connect ETIMEDOUT 192.168.1.172:9100","broker":"192.168.1.172:9100","clientId":"example-consumer","stack":"Error: connect ETIMEDOUT 192.168.1.172:9100\n    at Object._errnoException (util.js:992:11)\n    at _exceptionWithHostPort (util.js:1014:20)\n    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1186:14)"}
{"level":"ERROR","timestamp":"2019-02-07T22:24:20.080Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9094","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":20,"size":228}
{"level":"ERROR","timestamp":"2019-02-07T22:24:23.231Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 4)","broker":"192.168.1.172:9094","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":21,"size":261}
{"level":"INFO","timestamp":"2019-02-07T22:24:29.462Z","logger":"kafkajs","message":"[Runner] Consumer has joined the group","groupId":"test-group","memberId":"example-consumer-01cc906d-a0f7-47e4-a7b8-847178131aee","leaderId":"example-consumer-01cc906d-a0f7-47e4-a7b8-847178131aee","isLeader":true,"memberAssignment":{"topic-test":[2,5,4,1,3,0]},"groupProtocol":"RoundRobinAssigner","duration":12256}
```

--- a/src/consumer/__tests__/errorRecovery.spec.js
+++ b/src/consumer/__tests__/errorRecovery.spec.js
@@ -1,6 +1,7 @@
 const createProducer = require('../../producer')
 const createConsumer = require('../index')
 const { MemberMetadata, MemberAssignment } = require('../../consumer/assignerProtocol')
+const { KafkaJSError } = require('../../errors')
 
 const {
   secureRandom,
@@ -107,6 +108,51 @@ describe('Consumer', () => {
         }),
       },
     ])
+  })
+
+  it('recovers from crashes due to retriable errors', async () => {
+    consumer = createConsumer({
+      cluster,
+      groupId,
+      logger: newLogger(),
+      heartbeatInterval: 100,
+      maxWaitTimeInMs: 1,
+      maxBytesPerPartition: 180,
+      retry: {
+        retries: 0,
+      },
+    })
+    const crashListener = jest.fn()
+    consumer.on(consumer.events.CRASH, crashListener)
+
+    const error = new KafkaJSError(new Error('💣'), { retriable: true })
+
+    await consumer.connect()
+    await consumer.subscribe({ topic: topicName, fromBeginning: true })
+
+    const coordinator = await cluster.findGroupCoordinator({ groupId })
+    const original = coordinator.joinGroup
+    coordinator.joinGroup = async () => {
+      coordinator.joinGroup = original
+      throw error
+    }
+
+    const eachMessage = jest.fn()
+    await consumer.run({ eachMessage })
+
+    const key = secureRandom()
+    const message = { key: `key-${key}`, value: `value-${key}` }
+    await producer.send({ acks: 1, topic: topicName, messages: [message] })
+
+    await waitFor(() => crashListener.mock.calls.length > 0)
+    expect(crashListener).toHaveBeenCalledWith({
+      id: expect.any(Number),
+      timestamp: expect.any(Number),
+      type: 'consumer.crash',
+      payload: { error, groupId },
+    })
+
+    await expect(waitFor(() => eachMessage.mock.calls.length)).resolves.toBe(1)
   })
 
   describe('when eachMessage throws an error', () => {

--- a/src/consumer/index.js
+++ b/src/consumer/index.js
@@ -202,7 +202,7 @@ module.exports = ({
         groupId,
       })
 
-      if (e.name === 'KafkaJSNumberOfRetriesExceeded') {
+      if (e.name === 'KafkaJSNumberOfRetriesExceeded' || e.retriable === true) {
         logger.error(`Restarting the consumer in ${e.retryTime}ms`, {
           retryCount: e.retryCount,
           groupId,

